### ISSUE TITLE
perf(train): 兼容 LyCORIS Triton 与 FlashAttention

### ISSUE THREAD

**Issue by WalkingMeatAxolotl** [WalkingMeatAxolotl/AnimaLoraStudio#588]
## Parent

- Tracking roadmap: #567（R6）
- Depends on completed R5 measurement: #576
- Related stable-upgrade gate: #570（LoRA/LoHa 的本项工作不被其阻塞）

## 背景

Anima 在 BF16/FP16 autocast 下的 AdaLN/LayerNorm 归一化结果仍为 FP32。LyCORIS 4.0.0 的 Triton bypass 会把 LoRA/LoHa adapter 权重通过 `.to(x)` 对齐到该输入，并按 operand promotion 规则返回 FP32：只要任一 operand 为 FP32，输出即为 FP32。

这会把 attention 的 Q/K/V 提升为 FP32。FlashAttention 只接受 FP16/BF16，因此捕获异常后持续回退到 PyTorch SDPA。日志对相同 stage + shape 去重，所以只出现一次 warning 不代表只回退一次；LEAP 不是根因，但多次模型前向会放大性能和显存影响。

当前安全措施是让 `lycoris_backend=triton` 与 `attention_backend=flash_attn` 在配置层互斥。本 issue 记录未来恢复两者兼容并移除该限制的工作，不在当前安全修复中直接改变 kernel 或模型数值路径。

## 已验证事实

环境：RTX 5090、PyTorch 2.11.0+cu128、LyCORIS 4.0.0、flash-attn 2.8.3、BF16 autocast。

- 当前 Triton LoRA bypass：FP32 normalized activation → FP32 Q/K/V → FlashAttention 拒绝。
- 在进入 Triton bypass 前把 normalized activation 对齐到 active autocast dtype 后：
  - LoRA forward/backward 均成功且梯度有限；
  - LoHa forward/backward 均成功且梯度有限；
  - `[2, 3952, 16, 128]` 的 BF16 Q/K/V 可成功执行 FlashAttention；
  - 没有 SDPA fallback。
- 仅将 Triton 输出在末端 cast 为 BF16 虽能让 FlashAttention 接受，但仍保留 FP32 kernel 工作，性能收益不足。

代表性 LoRA delta 微基准（input `[2, 3952, 2048]`、rank 16，仅供方案比较，不代表端到端训练吞吐）：

| 路径 | 输出 dtype | forward + backward |
|---|---:|---:|
| Torch eager | BF16 | 0.851 ms |
| 当前 Triton | FP32 | 1.155 ms |
| Triton，输入对齐 BF16 | BF16 | 0.854 ms |
| Triton，仅输出 cast BF16 | BF16 | 1.120 ms |

LyCORIS 4.0.0 将 fused kernels 标为 early experimental；其 precision 文档明确记录 FP32 operand 导致 FP32 output。调查时上游 `main` 的 LoCon bypass 仍直接使用原始 `x`，未发现已合并的 autocast 边界修复。

上游参考：

- https://github.com/KohakuBlueleaf/LyCORIS/releases/tag/v4.0.0
- https://github.com/KohakuBlueleaf/LyCORIS/blob/v4.0.0/docs/kernels/precision.md

## 建议实现方向

优先评估在 Anima AdaLN normalized activation 边界按 active autocast dtype 做一次显式对齐，而不是在 FlashAttention 前或 Triton 输出后补 cast：

```python
normalized = layer_norm(x) * (1 + scale) + shift
if normalized.is_cuda and torch.is_autocast_enabled("cuda"):
    normalized = normalized.to(torch.get_autocast_dtype("cuda"))
```

需要覆盖普通 `forward` 和 `forward_tokens` / packed path 中的 self-attention、cross-attention 与 MLP。该位置允许 Q/K/V 复用一次转换，也与 eager `Linear` 在 autocast 下的实际输入 dtype 一致。

备选方案是在版本守卫的 LyCORIS patch 中，让 LoCon/LoHa bypass delta 使用 `org_forward(x).dtype`；它更通用，但可能为 Q/K/V 重复转换同一个 activation，并增加第三方类方法 patch 的维护成本。

## 验收标准

- [ ] BF16 与 FP16 autocast 下，Triton LoRA 的 Q/K/V 保持 active autocast dtype。
- [ ] Triton LoHa 同样保持正确 dtype。
- [ ] 普通 forward 与 packed / `forward_tokens` 路径均覆盖。
- [ ] LoRA/LoHa forward、input gradient 与 adapter parameter gradient 全部有限，并与 Torch backend 在约定容差内一致。
- [ ] `[2, 3952, 16, 128]` 代表性 shape 确认实际进入 FlashAttention，不再静默回退 SDPA。
- [ ] baseline sampling、普通训练及所有 LEAP variant 通过。
- [ ] 比较修复前后端到端峰值显存与训练吞吐，确认没有用末端 cast 掩盖 FP32 kernel 工作。
- [ ] 通过相关 pytest、route/config 横切测试、ruff，以及适用的前端 schema/UI 测试。
- [ ] 验证完成后移除 Triton 与 FlashAttention 的配置互斥及临时说明。
